### PR TITLE
Added the salon creation flow

### PR DESCRIPTION
**Pull request by mazemaze**
https://youtu.be/nJ8gmkC65mw 
スクリーンショット 2021-08-05 12 31 24

I added a function to add media files and upload them to both firestore and cloud storage.
And also, when an item is added to a list that is not on the screen, automatically scrolls to a position of the latest item.
Users can only delete items from the media previewer, however, I'm considering adding a delete button at the corner of each item in the media pane.
Added a function to check if none of the fields are empty which is what you've told me before.
I haven't added a function to change the order of items in the media pane. 
And when users tap the skip button, only users' email will be added into firestore so that they reach the home screen instead of the salon creation screen when they open the application after they skip the salon creation.

### CODE REVIEW

File: salon_creator/lib/app.dart [hunk 1]
```diff
@@ -32,7 +32,12 @@ class SalonCreatorApp extends StatelessWidget {
       home: StreamBuilder(
         stream: FirebaseAuth.instance.authStateChanges(),
         builder: (ctx, snapshot) {
-          if (snapshot.hasData) {
+          final User user = snapshot.data;
```
Comment (dragonfly91) [inline salon_creator/lib/app.dart:35]: What is the purpose of this change?

File: salon_creator/lib/models/salon.dart [hunk 2]
```diff
@@ -1,11 +1,13 @@
 class Salon {
+  String owner;
   String name;
-  String media;
+  List media;
```
Comment (dragonfly91) [inline salon_creator/lib/models/salon.dart]: Is this a `List<String>`?

File: salon_creator/lib/models/user_model.dart [hunk 3]
```diff
@@ -31,13 +31,13 @@ class UserModel {
   }
 
   UserModel.fromMap(Map<String, dynamic> map)
-      : assert(map['email'] != null),
-        assert(map['role'] != null),
-        email = map['email'],
+      : email = map['email'],
         role = RoleType.values[map['role']],
-        settings = UserSettings.fromMap(map['setting'].cast<String, dynamic>()),
+        settings = UserSettings.fromMap(map['setting']),
```
Comment (dragonfly91) [inline salon_creator/lib/models/user_model.dart]: Why is this change necessary?

File: salon_creator/lib/screens/login_screen.dart [hunk 4]
```diff
@@ -162,12 +162,11 @@ class _LoginScreenState extends State<LoginScreen> {
     final User user = FirebaseAuth.instance.currentUser;
 
     if (user != null) {
-      assert(!user.isAnonymous);
-      assert(await user.getIdToken() != null);
-
-      if (await dbHandler.getUser(user.email) != null) {
+      if (dbHandler.getUser(user.email) != null) {
```
Comment (dragonfly91) [inline salon_creator/lib/screens/login_screen.dart]: Why remove await?

File: salon_creator/lib/screens/salon_creation_screen.dart [hunk 5]
```diff
@@ -8,20 +20,529 @@ class SalonCreationScreen extends StatefulWidget {
...
+
+  ScrollController _scrollController = ScrollController();
+  TextEditingController categoriesController = TextEditingController();
+  TextEditingController descriptionController = TextEditingController();
+  TextEditingController priceController = TextEditingController();
+  TextEditingController salonNameController = TextEditingController();
+  XFile image;
+
+  void _updateScreenContext() {
+    var _hasProfileChanged = salonNameController.text.isNotEmpty &&
```
Comment (dragonfly91) [inline salon_creator/lib/screens/salon_creation_screen.dart]: I think these checks have to be OR checks. Profile is considered change if ANY of these become true.

File: salon_creator/lib/screens/salon_creation_screen.dart [hunk 6]
```diff
@@ -8,20 +20,529 @@ class SalonCreationScreen extends StatefulWidget {
...
+
     return Scaffold(
-      appBar: AppBar(
-        actions: [
-          TextButton(
-            onPressed: () {
-              Navigator.of(context).pushReplacementNamed('/home');
-            },
-            child: Text("スキップ"),
+      appBar: inProgress ? null : _buildAppBar(context),
```
Comment (dragonfly91) [inline salon_creator/lib/screens/salon_creation_screen.dart]:
```suggestion
      appBar: operationInProgress ? null : _buildAppBar(context),
```
Rename the variable for better readability?

File: salon_creator/lib/screens/salon_creation_screen.dart [hunk 7]
```diff
@@ -8,20 +20,529 @@ class SalonCreationScreen extends StatefulWidget {
...
+      isFilled = _hasProfileChanged;
+    });
+  }
+
+  _scrollToRight() {
+    _scrollController.jumpTo(_scrollController.position.maxScrollExtent);
+  }
+
   @override
   Widget build(BuildContext context) {
```
Comment (dragonfly91) [inline salon_creator/lib/screens/salon_creation_screen.dart:50]: Method too long

File: salon_creator/lib/screens/salon_creation_screen.dart [hunk 8]
```diff
@@ -8,20 +20,529 @@ class SalonCreationScreen extends StatefulWidget {
...
+                        ),
+                        _buildMediaPreviewerDivider(),
+                        SizedBox(
+                          height: 8,
+                        ),
+                        _buildMediaPreviewer(
+                            screenHeight, screenWidth, context),
+                        _buildMediaPane(context),
+                        _buildSalonDetailForms(screenWidth),
+                        isFilled ? Container() : Text("空欄の部分があります"),
```
Comment (dragonfly91) [inline salon_creator/lib/screens/salon_creation_screen.dart]:
```suggestion
                        isFormFilled ? Container() : Text("空欄の部分があります"),
```
Rename for better readability?

File: salon_creator/lib/screens/salon_creation_screen.dart [hunk 9]
```diff
@@ -8,20 +20,529 @@ class SalonCreationScreen extends StatefulWidget {
...
+  TextEditingController salonNameController = TextEditingController();
+  XFile image;
+
+  void _updateScreenContext() {
+    var _hasProfileChanged = salonNameController.text.isNotEmpty &&
+        categoriesController.text.isNotEmpty &&
+        descriptionController.text.isNotEmpty &&
+        priceController.text.isNotEmpty &&
+        _imageFiles.isNotEmpty;
+    print(salonNameController.text.isNotEmpty);
```
Comment (dragonfly91) [inline salon_creator/lib/screens/salon_creation_screen.dart]: Are these print statements necessary?

File: salon_creator/lib/screens/salon_creation_screen.dart [hunk 10]
```diff
@@ -8,20 +20,529 @@ class SalonCreationScreen extends StatefulWidget {
...
+                      ],
+                    ),
+                  ),
+                ),
+              ),
+            ),
+    );
+  }
+
+  Center _buildCreatingProgressScreen() {
```
Comment (dragonfly91) [inline salon_creator/lib/screens/salon_creation_screen.dart]:
```suggestion
  Widget _buildCreateInProgressScreen() {
```
Returned type should be Widget whenever possible
Rename the method name for better readability?

File: salon_creator/lib/screens/salon_creation_screen.dart [hunk 11]
```diff
@@ -8,20 +20,529 @@ class SalonCreationScreen extends StatefulWidget {
...
+          children: [
+            _buildMediaPaneAddButton(context),
+            _imageFiles != null ? _buildMediaPaneIconList() : Container(),
+          ],
+        ),
+      ),
+    );
+  }
+
+  Widget _buildMediaPaneIconList() {
```
Comment (dragonfly91) [inline salon_creator/lib/screens/salon_creation_screen.dart]:
```suggestion
  Widget _buildMediaPaneThumbnailList() {
```
For better readability.
Also, method too long

File: salon_creator/lib/screens/salon_creation_screen.dart [hunk 12]
```diff
@@ -8,20 +20,529 @@ class SalonCreationScreen extends StatefulWidget {
...
+        ),
+        onPressed: () {
+          showBottomSheet(context);
+        },
+      ),
+    );
+  }
+
+  Widget _buildMediaPreviewer(
+      double screenHeight, double screenWidth, BuildContext context) {
```
Comment (dragonfly91) [inline salon_creator/lib/screens/salon_creation_screen.dart]:
```suggestion
      double screenHeight, double screenWidth, BuildContext context,) {
```
nit

File: salon_creator/lib/screens/salon_creation_screen.dart [hunk 13]
```diff
@@ -8,20 +20,529 @@ class SalonCreationScreen extends StatefulWidget {
...
+        style: TextStyle(
+          fontSize: 18,
+        ),
+      ),
+      content: "選択されている画像・動画が削除されます",
+      leftButtonText: "削除する",
+      leftFunction: () {
+        setState(() {
+          _imageFiles.removeAt(index);
+          index -= 1;
```
Comment (dragonfly91) [inline salon_creator/lib/screens/salon_creation_screen.dart]:
```suggestion
          selectedImageIndex -= 1;
```
rename variable for better readability?

File: salon_creator/lib/screens/salon_creation_screen.dart [hunk 14]
```diff
@@ -8,20 +20,529 @@ class SalonCreationScreen extends StatefulWidget {
...
+        Navigator.of(context).pop();
+      },
+      rightButtonText: "取り消し",
+      rightFunction: () {
+        Navigator.of(context).pop();
+      },
+    );
+  }
+
+  void showBottomSheet(BuildContext context) async {
```
Comment (dragonfly91) [inline salon_creator/lib/screens/salon_creation_screen.dart]: Why not private?

File: salon_creator/lib/screens/salon_creation_screen.dart [hunk 15]
```diff
@@ -8,20 +20,529 @@ class SalonCreationScreen extends StatefulWidget {
...
+                title: Text("写真を選ぶ"),
+                onTap: () async {
+                  await _storeMediaFiles(context);
+                },
+              ),
+              Divider(
+                height: 0,
+              ),
+              ListTile(
+                title: Text("動画から選ぶ"),
```
Comment (dragonfly91) [inline salon_creator/lib/screens/salon_creation_screen.dart:358]: Are uploading videos not enabled for now?

File: salon_creator/lib/screens/salon_creation_screen.dart [hunk 16]
```diff
@@ -8,20 +20,529 @@ class SalonCreationScreen extends StatefulWidget {
...
+              width: 0.5,
+            ),
+          ),
+          hintText: "サロンの目的、発信内容等",
+        ),
+      ),
+    );
+  }
+
+  Future saveSalonDetail() async {
```
Comment (dragonfly91) [inline salon_creator/lib/screens/salon_creation_screen.dart]: Why not private?

File: salon_creator/lib/screens/salon_creation_screen.dart [hunk 17]
```diff
@@ -8,20 +20,529 @@ class SalonCreationScreen extends StatefulWidget {
...
+        index = _imageFiles.length - 1;
+      });
+      WidgetsBinding.instance.addPostFrameCallback((_) => _scrollToRight());
+      _updateScreenContext();
+    } catch (e) {
+      print(e);
+    }
+  }
+
+  Future _uploadImageAndGetUrl() async {
```
Comment (dragonfly91) [inline salon_creator/lib/screens/salon_creation_screen.dart]: I think this function is being used elsewhere as well. Why not move it to a class named StorageHandler (similar to DbHandler) and keep it there for common use?

**Review by dragonfly91 — CHANGES_REQUESTED** [review]
Gave some comments. Please address them.

File: salon_creator/lib/screens/salon_creation_screen.dart (hunk 5 above)
Comment (mazemaze) [reply salon_creator/lib/screens/salon_creation_screen.dart]: Do you mean that in salon creation users should be able to finish their salon creation even all of the forms are not filled?

File: salon_creator/lib/screens/salon_creation_screen.dart (hunk 15 above)
Comment (mazemaze) [reply salon_creator/lib/screens/salon_creation_screen.dart:358]: Yes, 'cause we decided that we'll do it separately.

File: salon_creator/lib/app.dart (hunk 1 above)
Comment (mazemaze) [reply salon_creator/lib/app.dart:35]: seems like it causes an error when the salon collection is not available. And I might forget to fix it.

File: salon_creator/lib/firebase/database.dart [hunk 18]
```diff
@@ -39,3 +44,27 @@ class DbHandler {
...
+
+  var imageFileRef;
+
+  StorageHandler() {
+    imageFileRef = FirebaseStorage.instance.ref().child(imageFilePath);
+  }
+
+  Future uploadMediaFilesAndGetUrls(List<XFile> mediaFiles) async {
+    var snapshot = imageFileRef;
+    List<String> donwloadUrl = [];
```
Comment (dragonfly91) [inline salon_creator/lib/firebase/database.dart]: downloadUrl

File: salon_creator/lib/models/user_model.dart [hunk 19]
```diff
@@ -30,14 +30,14 @@ class UserModel {
     return (role == RoleType.creator);
   }
 
+//Cast occurs error
```
Comment (dragonfly91) [inline salon_creator/lib/models/user_model.dart:33]: ??

**Review by dragonfly91 — COMMENTED** [review]
Left some comments

**Review by dragonfly91 — APPROVED** [review]
LGTM
